src/DcmTrajectoryGenerator.cpp

```diff
@@ -0,0 +1,652 @@
+    return false;
+  }
+
+  if(!nextSingleSupport->getDcmVel(doubleSupportEndTime, endVelocity)){
+    std::cerr << "[DCM TRAJECTORY GENERATOR] Error when the velocity of the DCM in the next SS phase is evaluated." <<std::endl;
+    return false;
+  }
+
+  // evaluate the new Double Support phase
+  std::shared_ptr<GeneralSupportTrajectory> currentDoubleSupport = std::make_shared<DoubleSupportTrajectory>(initPosition, initVelocity,
+													     endPosition, endVelocity,
+													     doubleSupportStartTime,
+													     doubleSupportEndTime);
+  // add the new Double Support phase
+  m_trajectory.push_back(currentDoubleSupport);
+  return true;
+}
+
+void DcmTrajectoryGenerator::getLastStepsTiming(double &singleSupportStartTime,
+						double &singleSupportEndTime,
+						double &singleSupportDuration,
+						double &doubleSupportEndTime,
+						double &singleSupportT0)
+{
+  doubleSupportEndTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+
+  singleSupportEndTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+  
+  singleSupportStartTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+ 
+  singleSupportT0  =  (singleSupportStartTime + m_phaseShift.back() * m_dT)/2;
+  singleSupportDuration = singleSupportEndTime  - singleSupportT0;
+}
+
+
+void DcmTrajectoryGenerator::getStepsTiming(double &singleSupportStartTime,
+					    double &singleSupportEndTime,
+					    double &singleSupportDuration,
+					    double &singleSupportT0)
+{
+  singleSupportEndTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+  
+  singleSupportStartTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+  
+  singleSupportT0  =  (singleSupportStartTime + m_phaseShift.back() * m_dT) / 2;
+
+  std::shared_ptr<GeneralSupportTrajectory> lastDoubleSupportPhase = m_trajectory.back();
+  std::pair<double, double> subtrajectoryDomain = lastDoubleSupportPhase->getTrajectoryDomain();
+  double doubleSupportEndTime = std::get<1>(subtrajectoryDomain);
+  singleSupportDuration = (singleSupportEndTime + doubleSupportEndTime) / 2 - singleSupportT0;
+}
+
+
+
+void DcmTrajectoryGenerator::getFirstDoubleSupportTiming(double &doubleSupportStartTime)
+{
+  doubleSupportStartTime = m_phaseShift.back() * m_dT;
+  m_phaseShift.pop_back();
+}
+
+
+bool DcmTrajectoryGenerator::generateDcmTrajectory(const std::vector<StepList::const_iterator> &orderedSteps,
+						   const StepList::const_iterator &firstStanceFoot,
+						   const iDynTree::Vector2 &initPosition,
+						   const iDynTree::Vector2 &initVelocity,
+						   const std::vector<size_t> &phaseShift)
+{
+
+  m_trajectoryDomain = std::make_pair(phaseShift.front(), phaseShift.back());
+  m_phaseShift = phaseShift;
+  m_orderedSteps = orderedSteps;
+
+  // reset the trajectory
+  m_trajectory.clear();
+  
+  // add the first stance foot at the beginning of the orderedStep vector
+  m_orderedSteps.insert(m_orderedSteps.begin(), firstStanceFoot);
```

Comment:
I understand this is actually a limitation of the first version of the planner. We may try to think about a better solution, also to avoid passing iterators to ``StepList``. Let's try to identify which are the requirements in this sense.